chalk-integration/src/lowering/env.rs

```diff
@@ -0,0 +1,271 @@
+///          // ----
+///          // |
+///          // addl_variable_kinds
+/// }
+/// ```
+#[derive(Debug, PartialEq, Eq)]
+pub struct AssociatedTyLookup {
+    pub id: chalk_ir::AssocTypeId<ChalkIr>,
+    pub addl_variable_kinds: Vec<chalk_ir::VariableKind<ChalkIr>>,
+}
+
+pub enum ApplyTypeLookup<'k> {
+    Parameter(&'k chalk_ir::WithKind<ChalkIr, BoundVar>),
+    Adt(AdtId<ChalkIr>),
+    FnDef(FnDefId<ChalkIr>),
+    Closure(ClosureId<ChalkIr>),
+    Opaque(OpaqueTyId<ChalkIr>),
+}
+
+impl Env<'_> {
+    pub fn interner(&self) -> &ChalkIr {
+        &ChalkIr
+    }
+
+    pub fn lookup_generic_arg(
+        &self,
+        name: &Identifier,
+    ) -> LowerResult<chalk_ir::GenericArg<ChalkIr>> {
+        let interner = self.interner();
+
+        let apply = |k: &TypeKind, type_name: chalk_ir::TypeName<ChalkIr>| {
+            if k.binders.len(interner) > 0 {
+                Err(RustIrError::IncorrectNumberOfTypeParameters {
+                    identifier: name.clone(),
+                    expected: k.binders.len(interner),
+                    actual: 0,
+                })
+            } else {
+                Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
+                    name: type_name,
+                    substitution: chalk_ir::Substitution::empty(interner),
+                })
+                .intern(interner)
+                .cast(interner))
+            }
+        };
+
+        match self.lookup_apply_type(name) {
+            Ok(ApplyTypeLookup::Parameter(p)) => {
+                let b = p.skip_kind();
+                Ok(match &p.kind {
+                    chalk_ir::VariableKind::Ty(_) => chalk_ir::TyData::BoundVar(*b)
+                        .intern(interner)
+                        .cast(interner),
+                    chalk_ir::VariableKind::Lifetime => chalk_ir::LifetimeData::BoundVar(*b)
+                        .intern(interner)
+                        .cast(interner),
+                    chalk_ir::VariableKind::Const(ty) => {
+                        b.to_const(interner, ty.clone()).cast(interner)
+                    }
+                })
+            }
+            Ok(ApplyTypeLookup::Adt(id)) => apply(self.adt_kind(id), chalk_ir::TypeName::Adt(id)),
+            Ok(ApplyTypeLookup::FnDef(id)) => {
+                apply(self.fn_def_kind(id), chalk_ir::TypeName::FnDef(id))
+            }
+            Ok(ApplyTypeLookup::Closure(id)) => {
+                apply(self.closure_kind(id), chalk_ir::TypeName::Closure(id))
+            }
+            Ok(ApplyTypeLookup::Opaque(id)) => Ok(chalk_ir::TyData::Alias(
+                chalk_ir::AliasTy::Opaque(chalk_ir::OpaqueTy {
+                    opaque_ty_id: id,
+                    substitution: chalk_ir::Substitution::empty(interner),
+                }),
+            )
+            .intern(interner)
+            .cast(interner)),
+            Err(_) => {
+                if let Some(id) = self.foreign_ty_ids.get(&name.str) {
```

Comment:
Is there anything stopping us from just making `lookup_apply_type` be able to return `Foreign`? It would clean this up and make it more uniform.

Reply:
What about the call in lowering `Ty` then? That doesn't need to match on `Foreign`

Reply:
I mean, we could either just `panic` there since we know we don't expect it.